pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import os
+from fnmatch import fnmatch
+
+from pootle.core.url_helpers import split_pootle_path
+from pootle_store.models import Store
+
+
+class VirtualFolderPathMatcher(object):
+
+    def __init__(self, vf):
+        self.vf = vf
+
+    def update_matching_stores(self):
+        existing_stores = self.vf.stores.all()
+        matching_stores = self.find_matching_stores()
+        to_add = []
+        to_delete = []
+        for store in matching_stores:
+            if store not in existing_stores:
+                to_add.append(store)
+        for store in existing_stores:
+            if store not in matching_stores:
+                to_delete.append(store.id)
+        if to_add:
+            self.vf.stores.add(*to_add)
+        if to_delete:
+            self.vf.stores.remove(
+                *self.vf.stores.filter(id__in=to_delete))
+        return to_add, to_delete
+
+    def add_store_if_matching(self, store):
+        pootle_dir_path = "".join(
+            ("/", ) + split_pootle_path(store.pootle_path)[2:])
+        no_add = (
+            not self.path_matches(pootle_dir_path)
+            or self.vf.stores.through.objects.filter(store_id=store.id).exists())
+        if no_add:
+            return False
+        self.vf.stores.add(store)
+        return True
+
+    @property
+    def filter_rules(self):
+        return (
+            "%s" % r.strip()
+            for r
+            in self.vf.filter_rules.split(","))
+
+    @property
+    def lang_code(self):
+        return (
+            self.vf.language
+            and self.vf.language.code)
+
+    @property
+    def proj_code(self):
+        return (
+            self.vf.project
+            and self.vf.project.code)
+
+    @property
+    def store_qs(self):
+        qs = Store.objects
+        if not self.vf.all_languages:
+            qs = qs.filter(
+                translation_project__language__in=self.vf.languages.all())
+        if not self.vf.all_projects:
+            qs = qs.filter(
+                translation_project__project__in=self.vf.projects.all())
+        qs = qs.filter(pootle_path__regex="^%s" % self.path)
+        return qs
+
+    @property
+    def path(self):
+        # TODO: filter to projects/languages if not all_*
```

Comment:
Then it is useless to keep this property? I wonder why you didn't chose to put full paths in the filters. I thought that was the plan, even if it was regex.